### ISSUE TITLE
Cache replacement policy

### ISSUE THREAD

**Issue by GaelVaroquaux** [joblib/joblib#85]
In the implementation of this feature, it is critical to avoid locks and race conditions. Thus all file system operation need to rely on atomic operations, such as appending a line to a file or moving a directory.
1. Log execution time, and disk space used during the computation
2. Accumulate usage statistics:
   This should be done in the form of an append-only text file where each usage is a line: appending a line in a file is atomic on most systems and thus we don't need a lock
3. Increment a count on the size of the cache dir
4. When the log file developed in 3 gets too big, move it to a new file (this is an atomic operation), and span a new process that make a summary statistic for each entry (done in a different process to avoid slowing down the master process).
5. Summary statistics should be a cost that only requires the last value to compute ('on line' computing to make an analogy with machine learning jargon). It should reflect the space used on disk (I'll denote it 'D'), the initial computation time ('T') and the usage pattern, i.e. the frequency of calls, time to last usage, function of the retrieval times, 't_i'.
   
   To design this statistic, here is my reasoning:
   - The statistic should have a 'forgetting factor': we don't care what the usage pattern was in the distant past, only the recent past is interesting. Thus we want a damping equation.
   - As a physicist, I think that we need to turn retrieval times into adimensional formats, i.e. compare them to a relevant time-scale, thus the damping factor should be as a function of 
     t_i / T
   - A simple damping equation is:
     C_n = sum_{i from 1 to n} exp((t_i - t) / T)
     where t is the current time
     The nice thing about this formula is that
     C_{n + 1} = (1 + exp(t_n - t_{n + 1}) C_n
     in other words it that be computed on-line
   
   Using the above 'memory' factor, one possible decision statistic is the following:
   
    S_n = T / D \* C_n
   
   C_n is bigger for cache items that get called more often.
